src/main.h

```diff
@@ -47,6 +47,28 @@ class CTxUndo;
 struct CNodeStateStats;
 class CTxInUndo;
 
+// Enforce 64-bit architecture requirement
+#if defined(_MSC_VER) || (defined(__INTEL_COMPILER) && defined(_WIN32))
+    #if defined(_M_X64)
+        #define BITNESS_64
+    #else 
+        #define BITNESS_32
+    #endif
+#elif defined(__clang__) || defined(__INTEL_COMPILER) || defined(__GNUC__)
+    #if defined(__x86_64)
+        #define BITNESS_64
+    #else
+        #define BITNESS_32
+    #endif
+#else
+    #error Cannot detect compiler or compiler is not supported
+#endif
+#if defined(BITNESS_32)
+#error "Zend is not supported on 32 bit architecture"
```

Comment:
> I do believe the whole point here is to detect BITNESS while making no assertion we do support any of these compilers. It is true we do not support MSVC but nothing prevents anybody from trying to build with it (fork the project and try): and at least BITNESS constraint will be honored.

I see your point, that was also my very first thought, but then 2 things made me prefer my suggestion: (1) we are adding lines of code to explicitly support this check for MSVC, and this is not consistent with the rest of our codebase and workflow, where never consider or check compatibility with MSVC explicitly. And (2), I read the message `or compiler is not supported` at line 67, combined with the explicit check for MSVC, as "we support the compilers above, including MSVC". I'd rather prevent this misunderstanding.

I also agree on the fact that someone might try to compile zend with MSVC (and we should also do that at some point), but given that we do not test it regularly and there might be many undocumented side effects (e.g. `NDEBUG`), I believe it makes sense to have this check fail out of the box. The brave user wanting to test MSVC can fix it very quickly, and we might even add a comment to facilitate that, in case.

> 
> > Same is true for architectures different from x86_64.
> 
> I'm not sure I understand the point unless you mean to change line 66 to `#if !defined(BITNESS_64)`

Yeah sorry that was not very clear. With the current code, if a brave user tries to compile for `ARM64` on `clang` or `g++`, we print the error message `Zend is not supported on 32 bit architecture`, which is misleading. I am suggesting to adjust the error message to account for this. For example, we might update the error message to something like `Zend only supports x86_64 architecture`.

Reply:
Modify or remove according to your needs